File: lib/src/main/java/it/scoppelletti/spaceship/app/AppExt.java
/*
 * Copyright (C) 2013-2016 Dario Scoppelletti, <http://www.scoppelletti.it/>.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package it.scoppelletti.spaceship.app;

import android.app.Activity;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.UiThread;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;
import android.text.TextUtils;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import it.scoppelletti.spaceship.ApplicationException;

/**
 * Application model operations.
 *
 * @since 1.0.0
 */
public final class AppExt {

    /**
     * Property indicating whether the user has manually opened the drawer at
     * least once.
     */
    public static final String PROP_LEARNED = "it.scoppelletti.spaceship.1";

    /**
     * Property reporting the title of an activity as a string resource ID.
     */
    public static final String PROP_TITLE = "it.scoppelletti.spaceship.2";

    /**
     * Tag of the {@code ConfirmDialogFragment} fragment.
     *
     * @see it.scoppelletti.spaceship.app.ConfirmDialogFragment
     */
    public static final String TAG_CONFIRMDIALOG =
            "it.scoppelletti.spaceship.1";

    /**
     * Tag of the {@code ExceptionDialogFragment} fragment.
     *
     * @see it.scoppelletti.spaceship.app.ExceptionDialogFragment
     */
    public static final String TAG_EXCEPTIONDIALOG =
            "it.scoppelletti.spaceship.2";

    /**
     * Private constructor for static class.
     */
    private AppExt() {
    }

    /**
     * Gets or creates a fragment.
     *
     * @param  activity      The activity hosting the fragment.
     * @param  fragmentClass Class of the fragment.
     * @param  tag           Tag of the fragment.
     * @param  <T>           Class of the fragment.
     * @return               The fragment.
     */
    @NonNull
    public static <T extends Fragment> T getOrCreateFragment(
            @NonNull Activity activity, @NonNull Class<T> fragmentClass,
            @NonNull String tag) {
        T fragment;
        FragmentManager fragmentMgr;

        if (activity == null) {
            throw new NullPointerException("Argument activity is null.");
        }
        if (fragmentClass == null) {
            throw new NullPointerException("Argument fragmentClass is null.");
        }
        if (TextUtils.isEmpty(tag)) {
            throw new NullPointerException("Argument tag is null.");
        }

        fragmentMgr = ((FragmentActivity) activity).getSupportFragmentManager();
        fragment = fragmentClass.cast(fragmentMgr.findFragmentByTag(tag));
        if (fragment != null) {
            return fragment;
        }

        try {
            fragment = fragmentClass.newInstance();
        } catch (IllegalAccessException ex) {
            throw new RuntimeException(ApplicationException.toString(ex), ex);
        } catch (InstantiationException ex) {
            throw new RuntimeException(ApplicationException.toString(ex), ex);
        }

        fragmentMgr.beginTransaction()
                .add(fragment, tag)
                .commit();

        return fragment;
    }

    /**
     * Hides the soft keyboard.
     *
     * @param activity The activity.
     */
    @UiThread
    public static void hideSoftKeyboard(@NonNull Activity activity) {
        View view;
        InputMethodManager inputMgr;

        if (activity == null) {
            throw new NullPointerException("Argument activity is null.");
        }

        view = activity.getCurrentFocus();
        if (view == null) {
            return;
        }

        inputMgr = (InputMethodManager) activity.getSystemService(
                Context.INPUT_METHOD_SERVICE);
        inputMgr.hideSoftInputFromWindow(view.getWindowToken(), 0);
    }
}

File: lib/src/main/kotlin/it/scoppelletti/spaceship/app/AppExt.kt
/*
 * Copyright (C) 2013-2023 Dario Scoppelletti, <http://www.scoppelletti.it/>.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package it.scoppelletti.spaceship.app

import android.content.pm.ApplicationInfo
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.os.Build
import androidx.annotation.RequiresApi

/**
 * Application model extensions.
 *
 * @since 1.0.0
 */
public object AppExt {

    /**
     * Tag of `AlertDialogFragment` fragment.
     *
     * @see it.scoppelletti.spaceship.app.AlertDialogFragment
     */
    public const val TAG_ALERTDIALOG: String = "it.scoppelletti.spaceship.1"

    /**
     * Tag of `ExceptionDialogFragment` fragment.
     *
     * @see it.scoppelletti.spaceship.app.ExceptionDialogFragment
     */
    public const val TAG_EXCEPTIONDIALOG: String = "it.scoppelletti.spaceship.2"

    /**
     * Tag of `BottomSheetDialogFragmentEx` fragment.
     *
     * @see it.scoppelletti.spaceship.app.BottomSheetDialogFragmentEx
     */
    public const val TAG_BOTTOMSHEETDIALOG: String =
        "it.scoppelletti.spaceship.3"

    /**
     * Property containing an item.
     */
    public const val PROP_ITEM: String = "it.scoppelletti.spaceship.1"

    /**
     * Property containing a message.
     */
    public const val PROP_MESSAGE: String = "it.scoppelletti.spaceship.2"

    /**
     * Property containing a result.
     */
    public const val PROP_RESULT: String = "it.scoppelletti.spaceship.3"

    /**
     * Returns information about an application.
     *
     * @param  packageManager PackageManager instance.
     * @param  packageName    Package name.
     * @param  flags          Info combination to retrieve.
     * @return                Application info.
     * @since                 1.1.0
     */
    @Throws(PackageManager.NameNotFoundException::class)
    public fun getApplicationInfo(
        packageManager: PackageManager,
        packageName: String,
        flags: Int
    ): ApplicationInfo = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU)
        getApplicationInfoTiramisu(packageManager, packageName, flags)
    else
        getApplicationInfoDeprecated(packageManager, packageName, flags)

    @RequiresApi(Build.VERSION_CODES.TIRAMISU)
    @Throws(PackageManager.NameNotFoundException::class)
    private fun getApplicationInfoTiramisu(
        packageManager: PackageManager,
        packageName: String,
        flags: Int
    ): ApplicationInfo = packageManager.getApplicationInfo(packageName,
        PackageManager.ApplicationInfoFlags.of(flags.toLong()))

    @Suppress("deprecation")
    @Throws(PackageManager.NameNotFoundException::class)
    private fun getApplicationInfoDeprecated(
        packageManager: PackageManager,
        packageName: String,
        flags: Int
    ): ApplicationInfo = packageManager.getApplicationInfo(packageName, flags)

    /**
     * Returns information about a package.
     *
     * @param  packageManager PackageManager instance.
     * @param  packageName    Package name.
     * @param  flags          Info combination to retrieve.
     * @return                Package info.
     * @since                 1.1.0
     */
    @Throws(PackageManager.NameNotFoundException::class)
    public fun getPackageInfo(
        packageManager: PackageManager,
        packageName: String,
        flags: Int
    ): PackageInfo = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU)
        getPackageInfoTiramisu(packageManager, packageName, flags)
    else
        getPackageInfoDeprecated(packageManager, packageName, flags)

    @RequiresApi(Build.VERSION_CODES.TIRAMISU)
    @Throws(PackageManager.NameNotFoundException::class)
    private fun getPackageInfoTiramisu(
        packageManager: PackageManager,
        packageName: String,
        flags: Int
    ): PackageInfo = packageManager.getPackageInfo(packageName,
        PackageManager.PackageInfoFlags.of(flags.toLong()))

    @Suppress("deprecation")
    @Throws(PackageManager.NameNotFoundException::class)
    private fun getPackageInfoDeprecated(
        packageManager: PackageManager,
        packageName: String,
        flags: Int
    ): PackageInfo = packageManager.getPackageInfo(packageName, flags)

    /**
     * Returns the version of a package.
     *
     * @param  packageInfo Package info.
     * @return             Version.
     * @since              1.1.0
     */
    public fun getVersion(packageInfo: PackageInfo): Long =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P)
            packageInfo.longVersionCode else getVersionDeprecated(packageInfo)

    @Suppress("deprecation")
    private fun getVersionDeprecated(packageInfo: PackageInfo): Long =
        packageInfo.versionCode.toLong()
}
